File: frontend/src/backend/userService.js
import {
  config,
  appFetch,
  setServiceToken,
  getServiceToken,
  removeServiceToken,
  setReauthenticationCallback,
} from "./appFetch";

export const login = (
  email,
  password,
  onSuccess,
  onErrors,
  reauthenticationCallback
) =>
  appFetch(
    "/users/login",
    config("POST", { email, password }),
    (authenticatedUser) => {
      setServiceToken(authenticatedUser.serviceToken);
      setReauthenticationCallback(reauthenticationCallback);
      onSuccess(authenticatedUser);
    },
    onErrors
  );

export const tryLoginFromServiceToken = (
  onSuccess,
  reauthenticationCallback
) => {
  const serviceToken = getServiceToken();

  if (!serviceToken) {
    onSuccess();
    return;
  }

  setReauthenticationCallback(reauthenticationCallback);

  appFetch(
    "/users/loginFromServiceToken",
    config("POST"),
    (authenticatedUser) => onSuccess(authenticatedUser),
    () => removeServiceToken()
  );
};

export const signUp = (user, onSuccess, onErrors, reauthenticationCallback) => {
  appFetch(
    "/users/signUp",
    config("POST", user),
    (authenticatedUser) => {
      setServiceToken(authenticatedUser.serviceToken);
      setReauthenticationCallback(reauthenticationCallback);
      onSuccess(authenticatedUser);
    },
    onErrors
  );
};

export const logout = () => removeServiceToken();

export const updateProfile = (user, onSuccess, onErrors) =>
  appFetch(`/users/${user.id}`, config("PUT", user), onSuccess, onErrors);

export const changePassword = (
  id,
  oldPassword,
  newPassword,
  onSuccess,
  onErrors
) =>
  appFetch(
    `/users/${id}/changePassword`,
    config("POST", { oldPassword, newPassword }),
    onSuccess,
    onErrors
  );

export const findClients = (keywords, page, size, onSuccess, onErrors) => {
  let path = `/users/clients?keywords=${keywords}&page=${page}&size=${size}`;

  appFetch(path, config("GET"), onSuccess, onErrors);
};

File: backend/src/main/java/es/udc/asiproject/service/UserService.java
package es.udc.asiproject.service;

import org.springframework.data.domain.Page;

import es.udc.asiproject.persistence.model.User;
import es.udc.asiproject.service.exceptions.DuplicateInstanceException;
import es.udc.asiproject.service.exceptions.IncorrectLoginException;
import es.udc.asiproject.service.exceptions.IncorrectPasswordException;
import es.udc.asiproject.service.exceptions.InstanceNotFoundException;

public interface UserService {
	void signUp(User user) throws DuplicateInstanceException;

	User login(String email, String password) throws IncorrectLoginException;

	User loginFromId(Long id) throws InstanceNotFoundException;

	User updateProfile(Long id, String firstName, String lastName) throws InstanceNotFoundException;

	void changePassword(Long id, String oldPassword, String newPassword)
			throws InstanceNotFoundException, IncorrectPasswordException;
	
	Page<User> findClients(Long id, String keywords, Integer pageNumber, Integer pageSize) throws InstanceNotFoundException;

	Page<User> findAllUsers(Integer pageNumber, Integer pageSize);

}
